### PR TITLE
Redux toolkit2

### PR DESCRIPTION
**Pull request by thyal**

### CODE REVIEW

File: src/lib/redux/store.ts [hunk 1]
```diff
@@ -1,59 +1,41 @@
...
 };
 
-export type RootState = ReturnType<typeof rootReducer>;
-export type AppDispatch = ReturnType<typeof createStore>["dispatch"];
-export type ThunkConfig = {
-    state: RootState;
-    dispatch: AppDispatch;
-    extra: { services: ReturnType<typeof createServices>; dispatchEvent: (event: CustomEvent) => void };
-};
+export const store = createStore({});
```
Comment (kevinwhereby) [inline src/lib/redux/store.ts:24]: this does cause one problem...
we can't inject services if we have one `store` instance to be accessed by both hooks, but we can't inject services from `useLocalMedia` because we don't know stuff like the room URL yet

Comment (kevinwhereby) [reply src/lib/redux/store.ts:24]: we _could_ have a dedicated store for local media, then a roomConnection store would subscribe to it to do stuff like send signal events on mute/unmute etc...

**Review by nandito — COMMENTED** [review]
The simplifications of the `RoomConnection.ts` are promising, like having in the `join()` fn this:

```js
this._store.dispatch(
            doAppJoin({
                roomName: this.roomName,
                roomKey: this._roomKey,
                displayName: this.displayName || "Guest",
                sdkVersion: sdkVersion || "unknown",
            })
        );
```
Instead of a series of different function calls. 👏 

Probably it's just because of the lack of my knowledge in the RTK world, but I find it really hard to follow what happens under the hood now. 😞 And as I got used to the currently published version I find it much easier to understand.

The approach is quite nice: having a similar frontend api like we have for pwa with react bundlers (`select..`, `do..`, `react..`). But the browser sdk is not a replacement of the pwa, it does not have such complex state (like tracking modals, notifications, and other ui states). Probably most of these states are going to be handled on the wrapper app side (I mean in the app that imports the browser-sdk).

Ideally the state management implementation should be clean, testable and abstract away the reused logic (boilerplates), but currently it looks like it's actually more logic with nested subscriptions/listeners/conditional dispatches. However, there's a good chance I just get distracted by the commented code and your changes made it simpler/smaller eventually. 🙈 

So after reading the code / _reviewing_ for more than an hour know I'm not fully convinced that this is the best solution we can choose. But I cannot offer an alternative and I really appreciate the effort here. Beginnings and changes can be hard initially and I'm ready to learn this tool if we start using it. 🙂

File: src/lib/RoomConnection.ts [hunk 2]
```diff
@@ -302,6 +313,46 @@ export default class RoomConnection extends TypedEventTarget {
...
+                                localParticipant: selectLocalParticipant(state) as LocalParticipant,
+                                remoteParticipants: selectRemoteParticipants(state),
+                                waitingParticipants: selectWaitingParticipants(state),
+                            },
+                        })
+                    );
+                }
+            }
+
+            if (state !== this._state) {
```
[the remark this answers is not in the document]
Comment (thyal) [reply src/lib/RoomConnection.ts:351]: True. I don't think we need this `_state` variable actually. We can just update the nested fields we care about when they change.

File: src/lib/RoomConnection.ts [hunk 3]
```diff
@@ -302,6 +313,46 @@ export default class RoomConnection extends TypedEventTarget {
         }: RoomConnectionOptions
     ) {
         super();
+        // redux
+        this._store = createStore({
+            injectServices: createServices(roomUrl),
```
[the remark this answers is not in the document]
Comment (thyal) [reply src/lib/RoomConnection.ts:318]: bad rebase I think, it should be passed to `createStore`.

File: src/lib/redux/store.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+export const store = createStore({});
+export type Store = typeof store;
+
+export const observeStore = <T>(store: Store, select: (state: RootState) => T, onChange: (result: T) => void) => {
+    let currentState: T;
+
+    function handleChange() {
+        const nextState = select(store.getState());
+        if (nextState !== currentState) {
```
[the remark this answers is not in the document]
Comment (kevinwhereby) [reply src/lib/redux/store.ts:32]: ah yeah you're right, would probably need a compare function here as well then

File: src/lib/redux/slices/localMedia.ts [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    },
+});
+startAppListening({
+    predicate: (_action, currentState, previousState) => {
+        const oldValue = selectCurrentMicrophoneDeviceId(previousState);
+        const newValue = selectCurrentMicrophoneDeviceId(currentState);
+        return oldValue !== newValue;
+    },
+    effect: (_action, { dispatch }) => {
+        dispatch(reactSetDevice({ audio: true, video: false }));
```
[the remark this answers is not in the document]
Comment (kevinwhereby) [reply src/lib/redux/slices/localMedia.ts:240]: the two listeners here (lines 2333 and 223) handle audio and video individually, like how `setCameraDevice` and `setMicrophoneDevice` both call `setDevice` in LocalMedia.ts currently

[the remark this answers is not in the document]
Comment (kevinwhereby) [reply src/lib/redux/slices/localMedia.ts:240]:
```typescript
    async setCameraDevice(deviceId: string) {
        this._currentCameraDeviceId = deviceId;
        await this._setDevice({ videoId: this._currentCameraDeviceId, audioId: false });
    }

    async setMicrophoneDevice(deviceId: string) {
        this._currentMicrophoneDeviceId = deviceId;
        await this._setDevice({ audioId: this._currentMicrophoneDeviceId, videoId: false });
    }
    /
    private async _setDevice({ audioId, videoId }: { audioId?: boolean | string; videoId?: boolean | string }) {
        const { replacedTracks } = await getStream(
            {
                ...this._getConstraintsOptions(),
                audioId,  // string or bool
                videoId, // string or bool
                type: "exact",
            },
            { replaceStream: this.stream }
        );
       // rest
       }
```

File: src/lib/RoomConnection.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (thyal) [reply src/lib/RoomConnection.ts:318]: ah no, not bad rebase. Kevin changed this to make it work with Local Media 😄 We will have two stores though so this will go back.

File: src/lib/redux/store.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinwhereby) [reply src/lib/redux/store.ts:32]: thinking about it some more, since we're using `createSelector` to create the `select` function, I _think_ the result is memo'd and so this comparison will work... but I'd have to double check

If that is the case, we'd probably want to enforce using a `createSelector` selector here with types

**Comment by kevinwhereby** [comment]
> But the browser sdk is not a replacement of the pwa

I think there's an argument that a lot of the pwaq logic we _do_ need at this layer (reconnection stuff, rtcstats stuff comes immediately to mind) could be copied over more easily if the implementation is similar, so even if the solution isn't the best suited, if it works and allows us to move faster with tried and tested code (rather than reimplementing), maybe it's a a tradeoff worth making?

File: src/lib/RoomConnection.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinwhereby) [reply src/lib/RoomConnection.ts:318]: my bad, I think I pushed the wrong branch this morning 😅

File: src/lib/redux/store.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinwhereby) [reply src/lib/redux/store.ts:32]: it works!
image
```typescript
export const observeStore = <T>(store: Store, select: (state: RootState) => T, onChange: (result: T) => void) => {
    let currentState: T;

    function handleChange() {
        const nextState = select(store.getState());
        console.log("trace state equality", nextState === currentState);
        if (nextState !== currentState) {
            currentState = nextState;
            onChange(currentState);
        }
    }

    const unsubscribe = store.subscribe(handleChange);
    handleChange();
    return unsubscribe;
};
```

File: src/lib/redux/slices/signalConnection.ts [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    doParticipantAudioEnabled,
+    doParticipantMetadataChanged,
+    doParticipantVideoEnabled,
+    doRoomJoined,
+} from "./room";
+import { doChatMessageReceived } from "./chat";
+import { doRoomConnectionStatusChanged } from "./roomConnection";
+import { doHandleCloudRecordingStarted, doHandleCloudRecordingStopped } from "./cloudRecording";
+
+const SIGNAL_BASE_URL = process.env["REACT_APP_SIGNAL_BASE_URL"] || "wss://signal.appearin.net";
```
Comment (nandito) [inline src/lib/redux/slices/signalConnection.ts:34]:
```suggestion
const SIGNAL_BASE_URL = process.env.REACT_APP_SIGNAL_BASE_URL || "wss://signal.appearin.net";
```

**Review by nandito — COMMENTED** [review]
Thanks for the update, now it's a bit easier to follow. There are still a few unusual concepts to learn (e.g. the listenerMiddleware). 🤪 But that's on me.

You made a nice progress, sadly there's quite a lot left 😬 and the ci is red. But I'm not against this approach anymore at least 😅 👏

File: src/lib/redux/store.ts (hunk 1 above)
Comment (havardholvik) [reply src/lib/redux/store.ts]: This comment might have come before our internal discussion. But I strongly believe LocalMedia should have its own store. Then RoomConnection could subscribe to it directly, or use the events which LocalMedia already dispatches.

File: src/lib/core/RoomConnection/redux/listenerMiddleware.ts [hunk 7]
```diff
@@ -11,3 +11,29 @@ export type AppStartListening = TypedStartListening<RootState, AppDispatch, Retu
...
+ * }));
+ * ```
+ * @param callback. The callback to be called on every action.
+ * @returns The unsubscribe function.
+ */
+export const createReactor = (
+    callback: ListenerEffect<AnyAction, RootState, AppDispatch, ReturnType<typeof createServices>>
+) => {
+    return startAppListening({
+        predicate: () => true,
```
Comment (kevinwhereby) [inline src/lib/core/RoomConnection/redux/listenerMiddleware.ts:36]: maybe I'm not understanding, but couldn't we prevent running every reactor for every action by moving the `if` logic from the effect to the predicate?

I.e. for the `deviceCredentials` reactor that calls `dispatch(doGetDeviceCredentials()`, we'd want the app listener to look like:
```typescript
startAppListening({
    predicate: () => {
        const wantsToJoin = selectAppWantsToJoin(getState());
        const deviceCredentials = selectDeviceCredentialsRaw(getState());
        return shouldFetchDeviceCredentials(wantsToJoin, deviceCredentials.isFetching, deviceCredentials.data)
    },
    effect: () =>  dispatch(doGetDeviceCredentials()),
})
```

Comment (kevinwhereby) [reply src/lib/core/RoomConnection/redux/listenerMiddleware.ts:36]: or even make `shouldFetchDeviceCredentials` a selector like:
```typescript
const shouldFetchDeviceCredentials = useSelector(selectAppWantsToJoin, selectDeviceCredentialsRaw, (wantsToJoin, deviceCredentials) => {
    if (wantsToJoin && !deviceCredentialsData && !isFetching) {
        return true;
    }
    return false;
}
```

so the reactor call could be simplified to:

```typescript
startAppListening({
    predicate: shouldFetchDeviceCredentials,
    effect: () =>  dispatch(doGetDeviceCredentials()),
})
```

but again, based on what Håvard said in standup I probably don't know the full story 😄

**Review by havardholvik — APPROVED** [review]
